### PR TITLE
[extractor/genius] Add extractor

### LINKED ISSUE #5209: Genius.com support request
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.10.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

United states 

### Example URLs

https://genius.com/videos/Vince-staples-breaks-down-the-meaning-of-when-sparks-fly
https://genius.com/videos/Breaking-down-drakes-certified-lover-boy-kanye-beef-way-2-sexy-cudi

### Provide a description that is worded well enough to be understood

yt-dlp can't extract audio nor video from the site 

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://genius.com/videos/Vince-staples-breaks-down-the-meaning-of-when-sparks-fly'] [debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8 [debug] yt-dlp version 2022.10.04 [4e0511f] (win32_exe) [debug] Python 3.8.10 (CPython 64bit) - Windows-10-10.0.19044-SP0 [debug] Checking exe version: ffmpeg -bsfs [debug] Checking exe version: ffprobe -bsfs [debug] exe versions: ffmpeg 2022-08-13-git-c469c3c3b1-full_build-www.gyan.dev (setts), ffprobe 2022-08-13-git-c469c3c3b1-full_build-www.gyan.dev [debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.09.24, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3 [debug] Proxy map: {} [debug] Loaded 1690 extractors [debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest Latest version: 2022.10.04, Current version: 2022.10.04 yt-dlp is up to date (2022.10.04) [debug] [generic] Extracting URL: https://genius.com/videos/Vince-staples-breaks-down-the-meaning-of-when-sparks-fly [generic] Vince-staples-breaks-down-the-meaning-of-when-sparks-fly: Downloading webpage WARNING: [generic] Falling back on generic information extractor [generic] Vince-staples-breaks-down-the-meaning-of-when-sparks-fly: Extracting information [debug] Looking for Brightcove embeds [debug] Looking for embeds ERROR: Unsupported URL: https://genius.com/videos/Vince-staples-breaks-down-the-meaning-of-when-sparks-fly Traceback (most recent call last):   File "yt_dlp\YoutubeDL.py", line 1477, in wrapper   File "yt_dlp\YoutubeDL.py", line 1553, in __extract_info   File "yt_dlp\extractor\common.py", line 672, in extract   File "yt_dlp\extractor\generic.py", line 3062, in _real_extract yt_dlp.utils.UnsupportedError: Unsupported URL: https://genius.com/videos/Vince-staples-breaks-down-the-meaning-of-when-sparks-fly
```

### PR DESCRIPTION
**Pull request by bashonly**
Adds support for genius.com video pages (atypical Brightcove embeds) and song lyrics pages (Youtube+Soundcloud embeds).

Closes #5209 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/genius.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        display_id = self._match_id(url)
+        webpage = self._download_webpage(url, display_id)
+
+        metadata = self._parse_json(self._html_search_regex(
+            r'<meta content="({[^"]*})"', webpage, 'metadata'), display_id)
+        video_id = traverse_obj(
+            metadata, ('video', 'provider_id'),
+            ('dfp_kv', lambda _, x: x['name'] == 'brightcove_video_id', 'values', 0), get_all=False)
+        if not video_id:
+            raise ExtractorError('Brightcove video id not found in webpage')
```
Comment (pukkandan) [inline yt_dlp/extractor/genius.py:55]: expected=True?

File: yt_dlp/extractor/genius.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            'thumbnail': r're:^https?://.*\.jpg$',
+        },
+    }]
+
+    def _real_extract(self, url):
+        display_id = self._match_id(url)
+        webpage = self._download_webpage(url, display_id)
+
+        metadata = self._parse_json(self._html_search_regex(
+            r'<meta content="({[^"]*})"', webpage, 'metadata'), display_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/genius.py:50]: `_search_json(... transform_source=unescapeHTML)`

File: yt_dlp/extractor/genius.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        },
+    }]
+
+    def _real_extract(self, url):
+        display_id = self._match_id(url)
+        webpage = self._download_webpage(url, display_id)
+
+        escaped_json = self._search_regex(
+            r'window\.__PRELOADED_STATE__\s*=\s*JSON\.parse\((.*)\);', webpage, 'escaped json', display_id)
+        unescaped_json = self._parse_json(escaped_json, display_id, transform_source=js_to_json)
```
Comment (pukkandan) [inline yt_dlp/extractor/genius.py:103]: note to self: see if this can be done with `_search_json`. Ref #5111

File: yt_dlp/extractor/genius.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            song_info, ('songPage', 'trackingData', lambda _, x: x['key'] == 'Primary Artist', 'value'),
+            get_all=False, default='unknown artist')
+        media = traverse_obj(
+            song_info, ('entities', 'songs', song_id, 'media'), expected_type=list, default=[])
+
+        entries = []
+        for m in media:
+            if m.get('type') in ('video', 'audio') and m.get('url'):
+                if m.get('provider') == 'spotify':
+                    continue
```
Comment (pukkandan) [inline yt_dlp/extractor/genius.py:122]: maybe this should give a message?

File: yt_dlp/extractor/genius.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                elif m.get('provider') == 'youtube':
+                    entries.append(self.url_result(m['url'], 'Youtube', song_id))
+                elif m.get('provider') == 'soundcloud':
+                    entries.append(self.url_result(m['url'], 'Soundcloud', song_id))
+                else:
+                    entries.append(self.url_result(m['url']))
```
Comment (pukkandan) [inline yt_dlp/extractor/genius.py:128]: Not worth it just to add `ie_key` imo
```suggestion
                else:
                    entries.append(self.url_result(m['url']))
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/genius.py (hunk 1 above)
Comment (bashonly) [reply yt_dlp/extractor/genius.py:55]: all of the current `/videos/` pages use brightcove embeds. I suppose that could change in the future but then we would want an issue reported in that case, no?

File: yt_dlp/extractor/genius.py (hunk 3 above)
Comment (bashonly) [reply yt_dlp/extractor/genius.py]: got it to work with `contains_pattern=r'\'{(?s:.+)}\''`

**Pull request merged**